hack/boilerplate/boilerplate.py

```diff
@@ -181,16 +185,91 @@ def get_regexs():
     regexs["go_build_constraints"] = re.compile(r"^(// \+build.*\n)+\n", re.MULTILINE)
     # strip #!.* from shell scripts
     regexs["shebang"] = re.compile(r"^(#!.*\n)\n*", re.MULTILINE)
+    # Search for generated files
+    regexs["generated"] = re.compile( 'DO NOT EDIT' )
     return regexs
 
+
+def is_generated_file(filename, regexs):
+    for d in skipped_ungenerated_files:
+ 	if d in filename:
+            return False
+    
+    try:
+        f = open(filename, 'r')
+    except Exception as exc:
+        print("Unable to open %s: %s" % (filename, exc), file=verbose_out)
+        return False
+
+    data = f.read()
+    f.close()
+
+    p = regexs["generated"]
+    found = p.search(data)
+
+    return found
+
+
+# If the file is automatically generated, then don't check the YEAR field.
+def generated_file_passes(filename, refs, regexs):
+    try:
+        f = open(filename, 'r')
+    except Exception as exc:
+        print("Unable to open %s: %s" % (filename, exc), file=verbose_out)
+        return False
+
+    data = f.read()
+    f.close()
+
+    extension = "generatego"
+    ref = refs[extension]
+
+    # remove build tags from the top of Go files
+    p = regexs["go_build_constraints"]
+    (data, found) = p.subn("", data, 1)
+
+    data = data.splitlines()
+
+    # if our test file is smaller than the reference it surely fails!
+    if len(ref) > len(data):
+        print('File %s smaller than reference (%d < %d)' %
+              (filename, len(data), len(ref)),
+              file=verbose_out)
+        return False
+
+    # trim our file to the same number of lines as the reference file
+    data = data[:len(ref)]
+
+    p = regexs["year"]
+    for d in data:
```

Comment:
isn't this the only actual difference to file_passes plus line 224 above? Can't we merge the two functions and react differently depending on whether is_generated_file return true or false? Also we would have read the file already, so we can avoid reading it twice this way.

Reply:
Yes, it has some differences in line 224 above, and also the debug log in line 246.
I have also considered merging them through the generate flag, but I'm afraid the code is too confusing. So separated them in two function.

Reply:
@sttts  and I can try to merge them now

Reply:
@sttts have merged and also squash change in 6 commits, thanks for review!